### ISSUE TITLE
Batch Normalization: Enable CUDA again

### ISSUE THREAD

**Issue by StefanoD** [fchollet/keras#6666]
The performance of BatchNormalization is really [slow](https://github.com/fchollet/keras/issues/6598). The reason is probably, because CUDA has been disabled for BatchNormalization.
[This](https://github.com/fchollet/keras/blob/master/keras/backend/theano_backend.py#L650) is the new code where CUDA isn't used.
The old [code](https://github.com/fchollet/keras/blob/master/keras/backend/theano_backend.py#L700) uses CUDA when available.

**Comment by StefanoD** [comment]
I misread the code...

**Comment by Hazarapet** [comment]
what does it mean? it calculates on the cpu?

**Comment by StefanoD** [comment]
I suppose so!

**Comment by Hazarapet** [comment]
Guys are there any news about this?

**Comment by the-moliver** [comment]
Keras uses `T.nnet.bn.batch_normalization_train` and `T.nnet.bn.batch_normalization_test`, both of which are CuDNN accelerated. If it's not working you need to update your Theano and CuDNN.

**Comment by StefanoD** [comment]
@the-moliver 
Could you prove this in code?
[This](https://github.com/Theano/Theano/blob/e79c4e4c83c5a4907ea7fddf073fd2d659df7486/theano/sandbox/cuda/dnn.py#L2926) is what the old functions calls. It calls CuDNN.

The new code uses a function in _theano/tensor/nnet/bn.py_.
[This](https://github.com/Theano/Theano/blob/e79c4e4c83c5a4907ea7fddf073fd2d659df7486/theano/tensor/nnet/bn.py#L42) functions documentation says:

> Also works on GPUs, but is not optimized using cuDNN.

But actually [this](https://github.com/Theano/Theano/blob/e79c4e4c83c5a4907ea7fddf073fd2d659df7486/theano/tensor/nnet/bn.py#L105) is called and the doc talks about cuDNN, but I can't see any invocation:

> epsilon : float
        Epsilon value used in the batch normalization formula. Minimum allowed
        value is 1e-5 (imposed by cuDNN).

The called AbstractBatchNormTrainGrad class just uses [Numpy](https://github.com/Theano/Theano/blob/e79c4e4c83c5a4907ea7fddf073fd2d659df7486/theano/tensor/nnet/bn.py#L603).

**Comment by StefanoD** [comment]
The [documentation](https://github.com/Theano/Theano/blob/e79c4e4c83c5a4907ea7fddf073fd2d659df7486/theano/tensor/nnet/bn.py#L162) of Theano says:
> If per-activation or spatial normalization is selected, this operation
    will use the cuDNN implementation.

Is this the case?

**Comment by the-moliver** [comment]
As far as I can tell. I believe the optimizers in Theano _should_ take care of selecting the appropriate implementation for your current software setup. If you can demonstrate it's not working with current drivers, CUDA, CuDNN, Theano, and gpuarray, then you should report it to the Theano devs. This thread may be helpful to you: https://github.com/Theano/Theano/pull/5190

**Comment by StefanoD** [comment]
Ok, thx. I close this, because I'm not an expert in Theano. Nevertheless: The performance is really disappointing. As [this](https://github.com/fchollet/keras/issues/6598) states: We need LRN in the upper layer as an alternative back again.

**Comment by the-moliver** [comment]
@StefanoD If I change the mnist_cnn.py example as below, (adding 3 BN layers) the time per epoch goes from 3s to 5s, which is about the performance hit percentage you should expect and I don't think is bad at all given the number of operations that are being done in each BN layer. If you're getting significantly worse performance, something else is wrong on your end.

```python

model = Sequential()
model.add(Conv2D(32, kernel_size=(3, 3),
                 activation='relu',
                 input_shape=input_shape))
model.add(BatchNormalization(axis=1))
model.add(Conv2D(64, (3, 3), activation='relu'))
model.add(BatchNormalization(axis=1))
model.add(MaxPooling2D(pool_size=(2, 2)))
model.add(Dropout(0.25))
model.add(Flatten())
model.add(Dense(128, activation='relu'))
model.add(BatchNormalization(axis=1))
model.add(Dropout(0.5))
model.add(Dense(num_classes, activation='softmax'))

model.compile(loss=keras.losses.categorical_crossentropy,
              optimizer=keras.optimizers.Adadelta(),
              metrics=['accuracy'])
```
Model summary:
```

Layer (type)                 Output Shape              Param #
=================================================================
conv2d_1 (Conv2D)            (None, 32, 26, 26)        320
_________________________________________________________________
batch_normalization_1 (Batch (None, 32, 26, 26)        128
_________________________________________________________________
conv2d_2 (Conv2D)            (None, 64, 24, 24)        18496
_________________________________________________________________
batch_normalization_2 (Batch (None, 64, 24, 24)        256
_________________________________________________________________
max_pooling2d_1 (MaxPooling2 (None, 64, 12, 12)        0
_________________________________________________________________
dropout_1 (Dropout)          (None, 64, 12, 12)        0
_________________________________________________________________
flatten_1 (Flatten)          (None, 9216)              0
_________________________________________________________________
dense_1 (Dense)              (None, 128)               1179776
_________________________________________________________________
batch_normalization_3 (Batch (None, 128)               512
_________________________________________________________________
dropout_2 (Dropout)          (None, 128)               0
_________________________________________________________________
dense_2 (Dense)              (None, 10)                1290
=================================================================
Total params: 1,200,778
Trainable params: 1,200,330
Non-trainable params: 448
_________________________________________________________________
```

**Comment by StefanoD** [comment]
@the-moliver 
I think, you don't have performance issues, because you have a small kernel size and very small images (28 x 28?). I have 72 x 272. I didn't do any measurements, but I would say the performance difference is a factor of 10 if I insert BatchNorm in the upper layers.

**Comment by the-moliver** [comment]
@StefanoD @Hazarapet No, I'm pretty sure you're either doing something wrong or have something set up wrong. I just tried resizing the MNIST images to 112x112 and reran the example: 51s / epoch without BN, 82s / epoch with the 3 BN layers as above. This amounts to nearly the same performance difference percentage as seen with 28x28 images. So if you're getting 10x worse performance something is definitely wrong on your end.

**Comment by Hazarapet** [comment]
BN layer's axis=1 for theano or tensorflow? I mean if the channel is the first, should I use BN with axis=1?

**Comment by StefanoD** [comment]
For both, when channel is first.

**Comment by Hazarapet** [comment]
great, that worked for me as well.

**Comment by the-moliver** [comment]
This is clearly stated in the docs btw: https://github.com/fchollet/keras/blob/master/keras/layers/normalization.py#L20

**Comment by StefanoD** [comment]
Yes, I have this from the docs.

**Comment by StefanoD** [comment]
But it could be done automatically.

**Comment by the-moliver** [comment]
Doesn't look like that's going to happen: https://github.com/fchollet/keras/issues/5328#issuecomment-303512067
